### PR TITLE
<497> Add anchors to main sections

### PR DESCRIPTION
**Pull request by federicobucchi**
### Motivation:

https://github.com/apple/swift-org-website/issues/497

### Modifications:

- `assets/stylesheets/_screen.scss` > added style to show `#` on `h2` hover
- `assets/javascripts/application.js` > adding `anchors` for each `h2` with `id` in the seen page

### Result:

Only targeting `h2` with `id` for now.

Screenshot 2024-03-23 at 9 48 26 PM

Screenshot 2024-03-23 at 9 49 00 PM

### CODE REVIEW

**Comment by federicobucchi** [comment]
@swift-ci please test

**Comment by alexandersandberg** [comment]
I believe this is a duplicate of #517?

**Review by kaishin — COMMENTED** [review]
I am inclined to vote for doing this using Jekyll if possible.

**Comment by federicobucchi** [comment]
@kaishin @alexandersandberg updated

**Comment by federicobucchi** [comment]
@daveverwer I added the `title` and the `chain link` icon.

I didn't move it to the left because I found that (in my research) that most of the anchors were placed on the left of the heading.

Screenshot 2024-04-24 at 10 38 27 AM

---

Screenshot 2024-04-24 at 10 38 38 AM

**Comment by federicobucchi** [comment]
@swift-ci please test

**Review by daveverwer — APPROVED** [review]
Looks great!

File: _plugins/convert-header.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+        if @options[:auto_ids] && !el_attribute['id']
+          el_attribute['id'] = generate_id(el.options[:raw_text])
+        end
+
+        el_attribute['class'] = "header-with-anchor"
+
+        @toc << [el.options[:level], el_attribute['id'], el.children] if el_attribute['id'] && in_toc?(el)
+        level = output_header_level(el.options[:level])
+        format_as_block_html("h#{level}", el_attribute, "#{inner(el, indent)} <a title=\"Permalink for #{inner(el, indent)}\ section" href=\"##{el_attribute['id']}\">
```
Comment (kaishin) [inline _plugins/convert-header.rb:15]:
```suggestion
        format_as_block_html("h#{level}", el_attribute, "#{inner(el, indent)} <a title=\"Permalink for #{inner(el, indent)}\ section\" href=\"##{el_attribute['id']}\">
```

**Review by kaishin — CHANGES_REQUESTED** [review]
The site wasn't building due to an unescaped character.

**Comment by kaishin** [comment]
@federicobucchi Beside the build issue above, the anchors are always visible on `h4`. They are also vertically misaligned and appear larger compared to the text.

We can perhaps use CSS to size the SVG using type-related units (`em`, `rem`) etc and if that doesn't work I wonder if we should revert to "#" as it would avoid this whole category of problems.

screenshot-2024-05-07

**Comment by federicobucchi** [comment]
@swift-ci please test

**Comment by federicobucchi** [comment]
@kaishin yeah that was a problem, but for now we want only `h2` so we shouldn't have the vertical align issue

Screenshot 2024-05-07 at 10 52 50 AM

Screenshot 2024-05-07 at 10 54 53 AM

**Comment by federicobucchi** [comment]
@swift-ci please test

**Comment by alexandersandberg** [comment]
Not sure if I'm missing anything but I don't see any anchors anywhere when running locally. I'm on the latest commit.

**Comment by federicobucchi** [comment]
@alexandersandberg I just tested, it works for me:
- Going to the page: `/getting-started/`
- Hover for example of "Using Swift"

**Comment by alexandersandberg** [comment]
> @alexandersandberg I just tested, it works for me:
> 
> * Going to the page: `/getting-started/`
> * Hover for example of "Using Swift"

Ok, so it worked if I first did `LC_ALL=en_us.UTF-8 bundle exec jekyll build` and then `... serve`. Probably due to some cache. As long as it works for people pulling the code for the first time it should be fine.

We need to add dark mode support though.

CleanShot 2024-05-09 at 08 16 43

**Comment by federicobucchi** [comment]
@alexandersandberg I have no idea of what you are experiencing unfortunately.

I am running `LC_ALL=en_us.UTF-8 bundle exec jekyll serve`

I have tried with `Auto` or forcing `Light` and `Dark` themes and all works correctly for me:

Screenshot 2024-05-09 at 3 35 49 PM

Screenshot 2024-05-09 at 3 35 42 PM

**Comment by alexandersandberg** [comment]
> @alexandersandberg I have no idea of what you are experiencing unfortunately.
> 
> I am running `LC_ALL=en_us.UTF-8 bundle exec jekyll serve`
> 
> I have tried with `Auto` or forcing `Light` and `Dark` themes and all works correctly for me:
> 
> Screenshot 2024-05-09 at 3 35 49 PM Screenshot 2024-05-09 at 3 35 42 PM

It works for me in Firefox, but not Safari.

**Comment by federicobucchi** [comment]
> > @alexandersandberg I have no idea of what you are experiencing unfortunately.
> > I am running `LC_ALL=en_us.UTF-8 bundle exec jekyll serve`
> > I have tried with `Auto` or forcing `Light` and `Dark` themes and all works correctly for me:
> > Screenshot 2024-05-09 at 3 35 49 PM Screenshot 2024-05-09 at 3 35 42 PM
> 
> It works for me in Firefox, but not Safari.

I tried with Safari, Firefox and Chrome.  All works for me.

@kaishin when you have the chance can you please try too?

If I get another approval, I will merge and we can try if @alexandersandberg experiences the same in prod (there is no risk since, at worst, @alexandersandberg wouldn't see the change)

**Review by kaishin — APPROVED** [review]
Thanks for addressing all the issues @federicobucchi! It is working fine on my side, in Safari and other browsers, in both light and dark modes.

**Pull request merged**